### PR TITLE
Rt blocked cancel ipc corres

### PR DESCRIPTION
**Pull request by mktnk3**
trying this out...

moving this PR from bitbucket to GitHub.

### CODE REVIEW

**Comment by lsf37** [comment]
I've put the label `MCS` on this -- I might try automating labelling for `rt` branch PRs later. I've also assigned the PR to @mktnk3 to indicate that Miki will be the person to merge it when she's happy with it.

Not sure we need to be very disciplined about this, but it enables filtering by assignee later.

**Comment by mktnk3** [comment]
Sounds reasonable to me.

> Don't worry about the failed style check, this should be fixed in the next merge from master, which I'll put up as soon as I get proof checks to work there.

Good to hear that. I was just wondering...

This PR needs a bit more cleanup, but in the meantime, yes, please use this to experiment with things as needed. I hope I can add more people on this when they are ready for it.

File: proof/refine/ARM/KHeap_R.thy [hunk 1]
```diff
@@ -2769,6 +2718,238 @@ lemma obj_relation_cuts_obj_bits:
...
+lemma Receive_or_Send_ep_at':
+  "\<lbrakk> st = BlockedOnReceive epPtr pl rp \<or> st = BlockedOnSend epPtr p1 p2 p3 p4;
+     valid_objs' s; st_tcb_at' ((=) st) t s\<rbrakk>
+       \<Longrightarrow> ep_at' epPtr s"
+  apply (drule (1) tcb_in_valid_state')
+  by (fastforce simp: obj_at'_def valid_tcb_state'_def)
+
+
+
+
```
Comment (mbrcknl) [inline proof/refine/ARM/KHeap_R.thy:2899]:
```suggestion

```

File: proof/refine/Move_R.thy [hunk 2]
```diff
@@ -250,4 +250,105 @@ lemma check_active_irq_invs_just_idle:
         and (\<lambda>s. 0 < domain_time s) and valid_domain_list \<rbrace>"
   by (wpsimp simp: check_active_irq_def ct_in_state_def)
 
+(* FIXME RT: Move? is this usable? *)
```
Comment (mbrcknl) [inline proof/refine/Move_R.thy:253]: I reckon keep it for now. I still have a suspicion there might be a generalisation, but haven't had any time to think about it.

**Comment by mktnk3** [comment]
I’ve added some new progress as a separate commit, including the state_refs_of cross lemma.

**Pull request merged**